### PR TITLE
Classification - models Experiments V1

### PR DESCRIPTION
**Pull request by NGimhana**
This PR contains classification models for ML experiments

### CODE REVIEW

File: ml_model/model/feature_selection/feature_selection.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+for feature in categorical_features:
+    for target in categorical_target:
+        tau,p = kendalltau(X[feature], y[target])
+        # write to csv
+        with open('ml_model/src/main/model/feature_selection/kendall_features.csv', "a") as csv_file:
+            writer = csv.DictWriter(csv_file, fieldnames=stat_header.keys())
+            writer.writerow({'input feature':feature, 'output_feature': target, 'tau': tau, 'p-value':p})
+## interprere the results
+# if p-value < 0.05, then the correlation is statistically significant. means that there is a relationship between the two variables.
+# filter based on above criteria and sort tau in descending order. now we have the most important features to predict the target variable.
```
Comment (ojcchar) [inline ml_model/model/feature_selection/feature_selection.py:39]: Fix: if |fau| > 0.1 and p-value < 0.05 then keep the feature

File: ml_model/model/classification/experiments.jsonl [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+{"experiment_id": "exp3-Kendals-BD50", "features": ["warnings_checker_framework", "TC (max)", "#conditionals (dft)" ,"Literals (Visual X)","Literals (Visual Y)","Literals (area)","Literals/identifiers (area)","Literals/keywords (area)"], "target": "BD50", "use_SMOTE" :false }
+{"experiment_id": "exp4-Kendals-PBU", "features": ["warnings_checker_framework", "MIDQ (min)" ,"NM (avg)","TC (avg)","Identifiers length (avg)","Line length (max)","Comments (Visual X)","Numbers (Visual X)"], "target": "PBU", "use_SMOTE" :true}
+{"experiment_id": "exp5-Kendals-ABU50", "features": ["warnings_checker_framework", "MIDQ (min)", "CICsyn (max)", "NM (avg)","TC (avg)","TC (max)","#comparisons (avg)","Identifiers length (avg)","#conditionals (avg)","#comparisons (dft)","#conditionals (dft)","Line length (dft)","Comments (Visual X)","#nested blocks (avg)","#parameters"], "target": "ABU50", "use_SMOTE" :true}
+{"experiment_id": "exp6-Kendals-BD50", "features": ["warnings_checker_framework", "TC (max)", "#conditionals (dft)" ,"Literals (Visual X)","Literals (Visual Y)","Literals (area)","Literals/identifiers (area)","Literals/keywords (area)"], "target": "BD50", "use_SMOTE" :true }
```
Comment (ojcchar) [inline ml_model/model/classification/experiments.jsonl:6]: We need to have extra experiments:
1. using all the features (code and total warnings).
2. using all the features (code and warnings separated by tool).

File: ml_model/model/feature_selection/feature_selection.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+## FILTER METHODS ##
+## write header
+with open('ml_model/src/main/model/feature_selection/kendall_features.csv', "w") as csv_file:
+    writer = csv.DictWriter(csv_file, fieldnames=stat_header.keys())
+    writer.writeheader()
+## Kendals Tau
+for feature in categorical_features:
+    for target in categorical_target:
+        tau,p = kendalltau(X[feature], y[target])
```
Comment (ojcchar) [inline ml_model/model/feature_selection/feature_selection.py:32]: Let's double check how NAs are treated

File: ml_model/model/classification/experiments.jsonl (hunk 2 above)
Comment (ojcchar) [reply ml_model/model/classification/experiments.jsonl:6]: Let's add more experiments by using the features selected by the Italians and warnings:
1. total
2. separated by tool

File: ml_model/model/results/compare_results.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
+import pandas as pd
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:1]: Add a description of what this file is doing and its purpose

File: ml_model/model/classification/main.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+                
+                ## for code features
+                # drop the frist 1 column
+                feature_X_c = full_dataset[experiment["features"]].iloc[:, 1:]
+                
+
+                ite = 1
+
+                ## keep track of the best model for each fold
+                best_hyperparams_c = set()
```
Comment (ojcchar) [inline ml_model/model/classification/main.py:313]: We just need one set of hyperparameters because there is no need/purpose for both

File: ml_model/model/results/compare_results.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+else:
+    SMOTE_TAG='NO_SMOTE'
+
+############################
+## FILTER OVERALL RESULTS ##
+############################
+## filter overall results
+overall_df = df.query("iteration == 'overall' and use_smote == " + str(use_SMOTE))
+## remove duplicate configs ##
+overall_df = overall_df.drop_duplicates(subset=['target','tp_c','tn_c','fp_c','fn_c','tp_cw','tn_cw','fp_cw','fn_cw','precision_c','recall_c','f1_c','precision_cw','recall_cw','f1_cw'], keep='first')
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:28]: This is probably not needed after we fix the problem with having two sets of best hyperparameters

File: ml_model/model/results/compare_results.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            metric_cw_mean = model_data[cw_metric].mean()
+            
+            ## median values
+            print(model_data.groupby(['model','target'])[metrics[metric]].median())
+            header[f'{c_metric} (median)'] = model_data.groupby(['model','target'])[c_metric].median().values[0]
+            header[f'{cw_metric} (median)'] = model_data.groupby(['model','target'])[cw_metric].median().values[0]
+            
+            ## mean values
+            print(model_data.groupby(['model','target'])[metrics[metric]].mean())
+            header[f'{c_metric} (mean)'] = model_data.groupby(['model','target'])[c_metric].mean().values[0]
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:91]: Maybe metric_c/cw_mean should be used here, because it is the same computation

File: ml_model/model/results/compare_results.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+
+## draw histogram - to check the normality for diff_f1, diff_precision, diff_recall
+columns = ['diff_precision','diff_recall','diff_f1', 'diff_auc']
+for column in columns:
+    hist = overall_df.hist(column=column)
+    plt.savefig(ROOT_PATH + 'results/histograms/final_hist_' + column +'_'+ str(SMOTE_TAG) + '.png')
+
+## draw histogram for all the input features in a single plot
+df_raw_features = pd.read_csv(ROOT_PATH + 'data/understandability_with_warnings.csv')
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:119]: I like the histograms in one file, but I don't know what to say. Nadeeshan, figure it out.

File: ml_model/model/results/compare_results.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+for model_name in models:
+    for target in targets:
+        model_data = overall_df.query("model == '" + model_name + "' and target == '" + target + "'")
+        print (model_data.shape[0])
+        for metric in metrics.keys():
+            x = model_data[metrics[metric][0]] ## code
+            y = model_data[metrics[metric][1]] ## code+warnings
+
+            ## 1. WILCOXON TEST ##
+            ## null hypothesis: There is no statistically significant difference between x and y
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:159]: small fixes: 
1. x=code and y=code+warnings
2. H0: x>y (fix the description: the performance (based on a metric) of the model using code features is greater than the performance when using code+warnings)

File: ml_model/model/results/compare_results.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            ## if p-value <= 0.05, we reject the null hypothesis. i.e. there is a statistically significant difference between x and y
+            ## Note: to handle ties (x=y), we use the zsplit method
+            _, wilcoxon_p = wilcoxon(x, y, alternative='greater', zero_method='zsplit')
+
+            ## write results to csv
+            tests['metric'] = metric
+            tests['model'] = model_name
+            tests['target'] = target
+            tests['wilcoxon_test(p-value)'] = wilcoxon_p
```
Comment (ojcchar) [inline ml_model/model/results/compare_results.py:168]: add the number of data points

Comment (ojcchar) [reply ml_model/model/results/compare_results.py:168]: add also the output of running the Cliff Delta's method:
https://github.com/neilernst/cliffsDelta

**Review by ojcchar — CHANGES_REQUESTED** [review]

File: ml_model/model/classification/main.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+from sklearn import naive_bayes
+
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import confusion_matrix, roc_curve, auc
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.neural_network import MLPClassifier
+from sklearn.svm import SVC, SVR
+
+## ignore all warnings comes from GridSearchCV
```
[the remark this answers is not in the document]
Comment (NGimhana) [reply ml_model/model/classification/main.py:19]: These are the warnings generated by models when they do not converge with a given set of hyperparameters.

File: ml_model/model/classification/main.py [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+    "use_smote": False
+}
+
+ROOT_PATH = "/Users/nadeeshan/Documents/Spring2023/ML-Experiments/complexity-verification-project/ml_model/model/"
+
+
+def model_initialisation(model_name, parameters):
+    LOGGER.info("Launching model: " + model_name + "...")
+
+    if model_name == "logistic_regression":
```
[the remark this answers is not in the document]
Comment (NGimhana) [reply ml_model/model/classification/main.py:76]: The hyperparameters that needs to tune are picked based on online resources, commented in the code

File: ml_model/model/classification/main.py [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+                ## StratifiedKFold
+                kf = StratifiedKFold(n_splits=folds, shuffle=True, random_state=42)
+                # kf = LeaveOneOut()
+                
+                ## for code + warning features
+                feature_X_cw = full_dataset[experiment["features"]]
+                
+                ## for code features
+                # drop the frist 1 column
+                feature_X_c = full_dataset[experiment["features"]].iloc[:, 1:]
```
[the remark this answers is not in the document]
Comment (NGimhana) [reply ml_model/model/classification/main.py:365]: Yes. It takes all the DF columns in the experiment["features"] array except the first one. The first one is the warning feature.

File: ml_model/model/feature_selection/feature_selection.py (hunk 3 above)
Comment (NGimhana) [reply ml_model/model/feature_selection/feature_selection.py]: kendalltau(X[feature1], X[feature2], nan_policy='omit') - This will ensure to perform pair-wise comparisons where both the data points have values (i.e not NAN)

File: ml_model/model/feature_selection/feature_selection.py (hunk 1 above)
Comment (NGimhana) [reply ml_model/model/feature_selection/feature_selection.py:270]: We now don't compare features and targets to get the correlation. So this is outdated
